Sp vl

aleien95 opened the pull request:
# What does this PR do?

Extends existing text sequence parallelism to support multimodal (image + text) training for SFT and DPO with padding removal optimization.

## Key Changes

- **Multimodal sequence parallelism for SFT/DPO**: Added support for vision-language models in both Supervised Fine-Tuning (SFT) and Direct Preference Optimization (DPO) training while maintaining backward compatibility with text-only training
- **Remove padding optimization**: Implemented dynamic padding removal to reduce memory usage and improve training efficiency
- **Enhanced parallel communication**: Updated tensor distribution patterns for multimodal data

Commit:
feat: add multimodal sequence parallelism support

Commit:
refactor: separate SFT/DPO training scripts and optimize data handling

- Remove data printing functions from SFT and DPO trainers for better performance
- Replace 360-example-vl.sh with separate SFT and DPO training scripts
- Add SFT visual-language demo dataset (data/sft-vl-demo/)
- Update dataset configuration to support new data structure

HaoshengZou commented on src/llamafactory/data/processors/sequence_parallel.py:
```diff
@@ -25,25 +41,61 @@ def pad_sequence(examples, data_args, tokenizer):
         for i in range(len(v)):
-            v[i].extend([pad_token_id] * (max_length - len(v[i])))
+             tmp_sp_len = max_input_ids_length_list[i] // model_args.sequence_parallel_size
+             closest_cutoff_len = int(tmp_sp_len + (8 - tmp_sp_len % 8)) * model_args.sequence_parallel_size
```
这里有保证每张卡上的序列pad到8的倍数吗？llama-factory每张卡都会 pad_to_multiple_of 8 (https://github.com/hiyouga/LLaMA-Factory/blob/main/src/llamafactory/train/dpo/workflow.py#L52)，序列并行不能被这个操作额外再pad一次，所以我们要保证SP切分后每张卡上已经是8的倍数

HaoshengZou commented on src/llamafactory/data/processors/sequence_parallel.py:
```diff
@@ -25,25 +41,61 @@ def pad_sequence(examples, data_args, tokenizer):
-            elif row is None:
+            #if k.endswith("attention_mask"):
+            #    chunks.extend([row] * model_args.sequence_parallel_size)
+            if row is None:
```
需要改前三行吗？

HaoshengZou commented on src/llamafactory/model/model_utils/sequence_parallel.py:
有更精简的写法吗？还是必须每个VL模型都改掉整个forward函数？可以先只看qwen2-vl qwen2.5-vl kimi-vl三个模型

HaoshengZou commented on src/llamafactory/train/sft/trainer.py:
```diff
@@ -97,11 +97,12 @@ def training_step(self, model, inputs, *args, **kwargs):
     @override
-    def _get_train_sampler(self):
+    def _get_train_sampler(self, *args, **kwargs):
+        train_dataset = args[0] if args else self.train_dataset
```
这里需要改吗？是为了和新版本transformers兼容？

aleien95 replied:
这里需要和新版本的transformers相兼容，新版本的llama-facotry也是这么改的

HaoshengZou reviewed (commented):
看起来应该不会有数据顺序的问题，要不再确认下数据顺序到底对不对？
大部分代码风格OK，主要就src/llamafactory/model/model_utils/sequence_parallel.py有没有可能精简一些，不要每个模型一个完整forward函数的patch。
不过还是优先保证正确性，先确认VL模型开SP后的数据顺序吧

Commit:
refactor: restructure multimodal model forward functions and optimize code style

- Add multimodal_forwards module to centrally manage multimodal model forward logic
- Extract and optimize forward function implementations for Qwen2 VL and Qwen2.5 VL
- Improve sequence_parallel related code structure
